### ISSUE TITLE
Wanpipe install broken on Debian 12, kernel 6.1

### ISSUE THREAD

**Issue by Famicoman** [InterLinked1/phreakscript#22]
Tried running phreakscript and getting my DAHDI device online today and noticed after install that wanpipe didn't seem to be build. When I try running manually I see that it does not seem to support kernel 6.1.x which is the default on my fresh installation.

```
        ----------------------------------------------------------
                  WANPIPE v7.0.34 Installation Script
            Copyright (c) 1995-2013, Sangoma Technologies Inc.
        ----------------------------------------------------------

WANPIPE INSTALLATION FOR DAHDI

You are about to install WANPIPE TDM Voice drivers
for Asterisk/Dahdi framework.

You will be prompted for path to DAHDI source.

Wanpipe drivers will compile into dahdi WITHOUT any
dahdi patching or need to recompile dahdi modules.

You must have Linux Kernel Headers along with
full development tools (i.e. GNU C compiler and utilities)
installed in order to be able to install this product.

If you have previoulsy installed WANPIPE, this release
will overrite/upgrade full release without the need to
uninstall first!

IMPORTANT:
It is always recommended to say YES to all options
prompted during the install!

Please visit: http://wiki.sangoma.com for more info.

Would you like to install WANPIPE now? [y] (y/n) y

        ----------------------------------------------------------
                  WANPIPE v7.0.34 Installation Script
            Copyright (c) 1995-2013, Sangoma Technologies Inc.
        ----------------------------------------------------------

Fixing file permissions...

Verifying files and fixing permissions ...Done
Checking for C developement tools ...(gcc) OK
Checking for C++ developement tools ...OK
Checking for Make utility ...OK
Checking for ncurses library ... OK
Checking for Perl developement tools ...OK
Checking for AWK ...OK
Checking for FLEX ...OK
Checking for Patch ...OK
Checking for libtermcap-devel...OK
Checking for bison...OK
Checking for libtool...OK


        ----------------------------------------------------------
                  WANPIPE v7.0.34 Installation Script
            Copyright (c) 1995-2013, Sangoma Technologies Inc.
        ----------------------------------------------------------


Installing WANPIPE Device Drivers: Linux KERNEL

To integrate WANPIPE Multi-Protocol Voice & WAN Router
modules into the Linux kernel, the kernel has to be
updated with latest wanpipe sources.  Install will only
modify existing wanpipe source that is already in the
Kernel.

IMPORTANT:
It is always recommended to say YES to all options
prompted during the install!


        ----------------------------------------------------------
                  WANPIPE v7.0.34 Installation Script
            Copyright (c) 1995-2013, Sangoma Technologies Inc.
        ----------------------------------------------------------


Please specify absolute path name of your linux headers/source directory

Press Enter for Default: /lib/modules/6.1.0-11-amd64/build

#>Setting linux directory to /lib/modules/6.1.0-11-amd64/build

./Setup: line 1350: [: : integer expression expected
./Setup: line 1350: [: : integer expression expected
./Setup: line 1350: [: : integer expression expected
./Setup: line 1360: [: : integer expression expected
./Setup: line 1364: [: : integer expression expected
./Setup: line 1367: [: : integer expression expected


Upgrading WANPIPE kernel documentation ...Done.


Installing WANPIPE include headers ...Done.

WANPIPE device drivers upgraded successfully!


        ----------------------------------------------------------
                  WANPIPE v7.0.34 Installation Script
            Copyright (c) 1995-2013, Sangoma Technologies Inc.
        ----------------------------------------------------------

WANPIPE KERNEL DRIVER COMPILATION

The next step in WANPIPE installation involves compiling
WANPIPE kernel modules.

This script will compile and install WANPIPE modules
into the currently running linux kernel.

For greater customization you will be prompted to
select which Protocol/Drivers you would like to
build into the WANPIPE kernel modules.

Wanpipe for Asterisk/Dahdi/Zaptel
        Default for Asterisk/Dahdi/Zaptel
Wanpipe for Wan Routing/API
        Default for Wan/IP Routing and Data API
Wanpipe for Asterisk SMG/SS7
        Default for Asterisk SS7
Wanpipe for TDM API
        Default for FreeSwitch and Voice API

Custom Compilation:
------------------
  Customise WANPIPE driver compilation to add only the
  protocols that you need.  This way one can reduce
  the size of the WANPIPE kernel drivers.

Refer to http://wiki.sangoma.com for more info
./Setup: line 1350: [: : integer expression expected
./Setup: line 1350: [: : integer expression expected
./Setup: line 1350: [: : integer expression expected
./Setup: line 1360: [: : integer expression expected
./Setup: line 1364: [: : integer expression expected
./Setup: line 1367: [: : integer expression expected


        ----------------------------------------------------------
                  WANPIPE v7.0.34 Installation Script
            Copyright (c) 1995-2013, Sangoma Technologies Inc.
        ----------------------------------------------------------


        Please Select Compilation Mode

1. WAN Protocols Support
       Protocols: Frame Relay, CHDLC, PPP, ATM, X25, ADSL, TDM API
       Default for: Wan Routing, Data & Voice API devel.

2. Asterisk/Dahdi Support
       Asterisk protocols: libpri (PRI,BRI), Analog (FXO/FXS), libss7 (SS7)
       Default for: Asterisk

3. Asterisk/Dahdi + WAN Protocol Support

4. TDM API (libsangoma)
       Protocols: TDM API (libsangoma) on AFT adapters:
       Default for: FreeSWITCH, Yate, Sunrise
                    Custom voice development

5. Custom Compilation Mode
       Specify protocols to be added into the WANPIPE
       kernel drivers.

6. Deprecated: SMG (BRI) (Asterisk SMG/BRI [Use Asterisk/Dahdi option 2 instead])

7. Deprecated: SMG (BRI) + Asterisk/Dahdi [Use Asterisk/Dahdi option 2 instead]


Please select (1-7) [Default: 1]:
        ----------------------------------------------------------
                  WANPIPE v7.0.34 Installation Script
            Copyright (c) 1995-2013, Sangoma Technologies Inc.
        ----------------------------------------------------------


Looking for zaptel/dahdi directory in /usr/src ...
-------------------------------------------
1 : /usr/src/dahdi-linux-3.2.0
------------------------------------------
------------------------------------------
m : Enter zaptel path manually

(ctrl-c to Exit)
Please select working zaptel directory [1-1][m]: 1

Enabling the TDM Voice Asterisk Support

Dahdi HW HDLC Support Detected: Enabling DCHAN Feature
Native Dahdi HW HDLC Support Detected - No patch required
Dahdi source unmodified

Checking for UDEV Zaptel compatibility...     Dahdi installed, no need to add UDEV rules
Enabling the AFT TE1 Support

Checking for SMP support ...Enabled.

Checking current processor type ...x86_64

./Setup: line 2654: [: -eq: unary operator expected
ERROR: Unsupported Kernel Version 6.1.0-11-amd64
root@kabelsalat:/usr/src/wanpipe-7.0.34#
```

**Comment by InterLinked1** [comment]
Do you need wanpipe? It's only necessary if you have a Sangoma card.

This duplicates #3, Sangoma is aware of the issue but has not released a fix yet, to my knowledge. However, if you report it to them that will help as they use that to determine how many people are affected!

**Comment by Famicoman** [comment]
Unfortunately for me I do need it based on what I am seeing when querying for hardware:
```
# dahdi_hardware
pci:0000:03:04.0     wanpipe-     1923:0040 Sangoma Technologies Corp. A200/Remora FXO/FXS Analog AFT card
```

Two questions:
1) Do we know the highest kernel supported? While annoying, I'm not opposed to downgrading my system.
2) What is the best way to contact Sangoma about this?

**Comment by InterLinked1** [comment]
> Unfortunately for me I do need it based on what I am seeing when querying for hardware:
> 
> ```
> # dahdi_hardware
> pci:0000:03:04.0     wanpipe-     1923:0040 Sangoma Technologies Corp. A200/Remora FXO/FXS Analog AFT card
> ```
> 
> Two questions:
> 
> 1. Do we know the highest kernel supported? While annoying, I'm not opposed to downgrading my system.

Ah, yeah it looks like you have a Sangoma card. They do require wanpipe and people do tend to have more issues with those.

This isn't a kernel issue, it's an issue with the script with Debian, but it works with Ubuntu. See the explanation about this from Dave in the linked issue.

> 2. What is the best way to contact Sangoma about this?

Two places - first, through sales, here: https://www.sangoma.com/contact-us/contact-sales/

Report the problem and how this impacts you. Sangoma takes sales more seriously than support.

Then, telephony cards, here: https://www.sangoma.com/support/

If you're not opposed to it and you want it to "just work" for now, you might also try running it on Ubuntu, or seeing if Dave's comments are of any help to you on making it work for Debian.

**Comment by Famicoman** [comment]
I do think this differs slightly from #3 now that I've tried to hardcode the proper directory based on what the Makefile points to. It appears that there is no logic in the Makefile for kernel 6.x and upon duplicating the 5.x logic I get a bit closer but it still fails.
It looks like #3 was attempting to compile with a 5.x kernel and the script explicity supports it, it just couldn't parse the proper values.

```
        ----------------------------------------------------------
                  WANPIPE v7.0.34 Installation Script
            Copyright (c) 1995-2013, Sangoma Technologies Inc.
        ----------------------------------------------------------


Looking for zaptel/dahdi directory in /usr/src ...
-------------------------------------------
1 : /usr/src/dahdi-linux-3.2.0
------------------------------------------
------------------------------------------
m : Enter zaptel path manually

(ctrl-c to Exit)
Please select working zaptel directory [1-1][m]: 1

Enabling the TDM Voice Asterisk Support

Dahdi HW HDLC Support Detected: Enabling DCHAN Feature
Native Dahdi HW HDLC Support Detected - No patch required
Dahdi source unmodified

Checking for UDEV Zaptel compatibility...     Dahdi installed, no need to add UDEV rules
Enabling the AFT TE1 Support

Checking for SMP support ...Enabled.

Checking current processor type ...x86_64

Checking fo regparm: Disabled


--------------------------------------------------
CFLAGS: gcc -Wp,-MD,.wanpipe.o.d -nostdinc -iwithprefix include -D__LINUX__ -Dlinux -D__KERNEL__ -I/usr/include/wanpipe -I/lib/modules/6.1.0-11-amd64/build/include  -DMODULE  make: Entering directory '/usr/src/linux-headers-6.1.0-11-amd64'
make: Leaving directory '/usr/src/linux-headers-6.1.0-11-amd64'
--------------------------------------------------

Compiling General WANPIPE Driver for 6.X.X Kernel .
---------------------------------------------


WANPIPE DRIVER COMPILE LOG
Mon Aug 14 01:24:30 PM EDT 2023
-----------------------------------
make -C /lib/modules/6.1.0-11-amd64/build SUBDIRS=/usr/src/wanpipe-7.0.34/kdrvtmp CC=gcc KBUILD_VERBOSE=1 modules
make[1]: Entering directory '/usr/src/linux-headers-6.1.0-11-amd64'
make -f /usr/src/linux-headers-6.1.0-11-common/scripts/Makefile.build obj=/usr/src/wanpipe-7.0.34/kdrvtmp need-builtin=1 need-modorder=1
  printf '%s
'   sdladrv_src.o sdladrv_fe.o sdladrv_utils.o wanpipe_cdev_linux.o wanpipe_logger.o sdladrv_usb.o | awk '!x[$0]++ { print("/usr/src/wanpipe-7.0.34/kdrvtmp/"$0) }' > /usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv.mod
  gcc -Wp,-MMD,/usr/src/wanpipe-7.0.34/kdrvtmp/.sdladrv_src.o.d -nostdinc -I/usr/src/linux-headers-6.1.0-11-common/arch/x86/include -I./arch/x86/include/generated -I/usr/src/linux-headers-6.1.0-11-common/include -I./include -I/usr/src/linux-headers-6.1.0-11-common/arch/x86/include/uapi -I./arch/x86/include/generated/uapi -I/usr/src/linux-headers-6.1.0-11-common/include/uapi -I./include/generated/uapi -include /usr/src/linux-headers-6.1.0-11-common/include/linux/compiler-version.h -include /usr/src/linux-headers-6.1.0-11-common/include/linux/kconfig.h -include /usr/src/linux-headers-6.1.0-11-common/include/linux/compiler_types.h -D__KERNEL__ -fmacro-prefix-map=/usr/src/linux-headers-6.1.0-11-common/= -Wall -Wundef -Werror=strict-prototypes -Wno-trigraphs -fno-strict-aliasing -fno-common -fshort-wchar -fno-PIE -Werror=implicit-function-declaration -Werror=implicit-int -Werror=return-type -Wno-format-security -std=gnu11 -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx -fcf-protection=none -m64 -falign-jumps=1 -falign-loops=1 -mno-80387 -mno-fp-ret-in-387 -mpreferred-stack-boundary=3 -mskip-rax-setup -mtune=generic -mno-red-zone -mcmodel=kernel -Wno-sign-compare -fno-asynchronous-unwind-tables -mindirect-branch=thunk-extern -mindirect-branch-register -mindirect-branch-cs-prefix -mfunction-return=thunk-extern -fno-jump-tables -mharden-sls=all -fno-delete-null-pointer-checks -Wno-frame-address -Wno-format-truncation -Wno-format-overflow -Wno-address-of-packed-member -O2 -fno-allow-store-data-races -Wframe-larger-than=2048 -fstack-protector-strong -Wno-main -Wno-unused-but-set-variable -Wno-unused-const-variable -Wno-dangling-pointer -ftrivial-auto-var-init=zero -fno-stack-clash-protection -pg -mrecord-mcount -mfentry -DCC_USING_FENTRY -Wdeclaration-after-statement -Wvla -Wno-pointer-sign -Wcast-function-type -Wno-stringop-truncation -Wno-stringop-overflow -Wno-restrict -Wno-maybe-uninitialized -Wno-array-bounds -Wno-alloc-size-larger-than -Wimplicit-fallthrough=5 -fno-strict-overflow -fno-stack-check -fconserve-stack -Werror=date-time -Werror=incompatible-pointer-types -Werror=designated-init -Wno-packed-not-aligned -g -D__LINUX__ -DCONFIG_PRODUCT_WANPIPE_BASE -DCONFIG_PRODUCT_WANPIPE_TDM_VOICE -DCONFIG_PRODUCT_WANPIPE_TDM_VOICE_DCHAN -DDAHDI_32 -DCONFIG_PRODUCT_WANPIPE_TDM_VOICE_DCHAN_ZAPTEL -DDAHDI_ISSUES -DCONFIG_PRODUCT_WANPIPE_AFT -DCONFIG_PRODUCT_WANPIPE_AFT_CORE -DCONFIG_PRODUCT_WANPIPE_AFT_TE1 -DCONFIG_PRODUCT_WANPIPE_AFT_56K -DCONFIG_PRODUCT_WANPIPE_AFT_RM -DCONFIG_PRODUCT_WANPIPE_CODEC_SLINEAR_LAW -DCONFIG_PRODUCT_WANPIPE_AFT_BRI -DCONFIG_PRODUCT_WANPIPE_AFT_SERIAL -DCONFIG_PRODUCT_WANPIPE_AFT_A600 -DCONFIG_PRODUCT_WANPIPE_AFT_B601 -DCONFIG_PRODUCT_WANPIPE_AFT_B800 -DCONFIG_PRODUCT_WANPIPE_AFT_A700 -DCONFIG_PRODUCT_WANPIPE_USB -DWANPIPE_USE_I_PRIVATE -DKERN_PROC_PDE_FEATURE=0 -DKERN_PROC_CREATE=0 -DKERN_SK_FOR_NODE_FEATURE=0 -DSOCK_NO_OPT=0 -DKERN_NETIF_TRANS_UPDATE=1 -DKERN_NDO_CHANGE_MTU_RH74=0 -DKERN_DEV_CHG_FLAG_UPDATE=1 -DKERN_NDO_TIMEOUT_UPDATE=1 -DKERN_MEMCPY_FROM_MSG=1 -DKERN_MEMCPY_TO_MSG=1 -DKERN_CLASS_DEV_GROUPS=1 -DKERN_TIMER_SETUP=1 -DKERN_TIMER_32BIT=1 -DKERN_REFCNT_UPDATE=1 -DKERN_SIG_UPDATE=1 -DKERN_CLASS_PROTO_OPS_UPDATE=0 -DKERN_DO_GET_TIME=0 -DKERN_GET_KENEL_DS=0 -DKERN_MODPOST_STATIC_ERR=0 -DCONFIG_WANPIPE_HWEC -I//usr/include/wanpipe -I/usr/include/wanpipe/oct6100_api -I/usr/include/wanpipe/oct6100_api/include -I/usr/include/wanpipe/oct6100_api/include/oct6100api  -DMODULE  -DKBUILD_BASENAME='"sdladrv_src"' -DKBUILD_MODNAME='"sdladrv"' -D__KBUILD_MODNAME=kmod_sdladrv -c -o /usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.o /usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c   ; ./tools/objtool/objtool --hacks=jump_label --hacks=noinstr --orc --retpoline --rethunk --sls --static-call --uaccess   --module /usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.o
In file included from //usr/include/wanpipe/wanpipe_defines.h:30,
                 from /usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:126:
//usr/include/wanpipe/wanpipe_kernel.h: In function ‘WP_PDE_DATA’:
//usr/include/wanpipe/wanpipe_kernel.h:295:36: error: implicit declaration of function ‘PDE’ [-Werror=implicit-function-declaration]
  295 |         struct proc_dir_entry* pde=PDE(inode);
      |                                    ^~~
//usr/include/wanpipe/wanpipe_kernel.h:295:36: warning: initialization of ‘struct proc_dir_entry *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
//usr/include/wanpipe/wanpipe_kernel.h:297:27: error: invalid use of undefined type ‘struct proc_dir_entry’
  297 |                 return pde->data;
      |                           ^~
In file included from /usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:128:
//usr/include/wanpipe/wanpipe_common.h: In function ‘wan_virt2bus’:
//usr/include/wanpipe/wanpipe_common.h:800:16: error: implicit declaration of function ‘virt_to_bus’; did you mean ‘virt_to_phys’? [-Werror=implicit-function-declaration]
  800 |         return virt_to_bus(ptr);
      |                ^~~~~~~~~~~
      |                virt_to_phys
//usr/include/wanpipe/wanpipe_common.h: In function ‘wan_bus2virt’:
//usr/include/wanpipe/wanpipe_common.h:819:16: error: implicit declaration of function ‘bus_to_virt’; did you mean ‘phys_to_virt’? [-Werror=implicit-function-declaration]
  819 |         return bus_to_virt(virt_addr);
      |                ^~~~~~~~~~~
      |                phys_to_virt
//usr/include/wanpipe/wanpipe_common.h:819:16: warning: returning ‘int’ from a function with return type ‘long unsigned int *’ makes pointer from integer without a cast [-Wint-conversion]
  819 |         return bus_to_virt(virt_addr);
      |                ^~~~~~~~~~~~~~~~~~~~~~
//usr/include/wanpipe/wanpipe_common.h: In function ‘wan_dma_alloc_org’:
//usr/include/wanpipe/wanpipe_common.h:905:28: error: implicit declaration of function ‘pci_alloc_consistent’ [-Werror=implicit-function-declaration]
  905 |         dma_descr->vAddr = pci_alloc_consistent(NULL,
      |                            ^~~~~~~~~~~~~~~~~~~~
//usr/include/wanpipe/wanpipe_common.h:905:26: warning: assignment to ‘long unsigned int *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  905 |         dma_descr->vAddr = pci_alloc_consistent(NULL,
      |                          ^
//usr/include/wanpipe/wanpipe_common.h: In function ‘wan_dma_free_org’:
//usr/include/wanpipe/wanpipe_common.h:951:9: error: implicit declaration of function ‘pci_free_consistent’ [-Werror=implicit-function-declaration]
  951 |         pci_free_consistent(NULL, dma_descr->max_length,dma_descr->vAddr,dma_descr->pAddr);
      |         ^~~~~~~~~~~~~~~~~~~
In file included from //usr/include/wanpipe/aft_core_user.h:14,
                 from //usr/include/wanpipe/wanpipe_api.h:39,
                 from //usr/include/wanpipe/wanpipe.h:46,
                 from /usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:129:
//usr/include/wanpipe/wanpipe_wanrouter.h: At top level:
//usr/include/wanpipe/wanpipe_wanrouter.h:339:9: error: unknown type name ‘write_proc_t’
  339 |         write_proc_t*   set_dev_config;
      |         ^~~~~~~~~~~~
//usr/include/wanpipe/wanpipe_wanrouter.h:340:9: error: unknown type name ‘write_proc_t’
  340 |         write_proc_t*   set_if_info;
      |         ^~~~~~~~~~~~
In file included from /usr/src/linux-headers-6.1.0-11-common/include/linux/byteorder/little_endian.h:5,
                 from /usr/src/linux-headers-6.1.0-11-common/arch/x86/include/uapi/asm/byteorder.h:5,
                 from /usr/src/linux-headers-6.1.0-11-common/arch/x86/include/asm/orc_types.h:43,
                 from /usr/src/linux-headers-6.1.0-11-common/arch/x86/include/asm/unwind_hints.h:6,
                 from /usr/src/linux-headers-6.1.0-11-common/arch/x86/include/asm/nospec-branch.h:13,
                 from /usr/src/linux-headers-6.1.0-11-common/arch/x86/include/asm/paravirt_types.h:40,
                 from /usr/src/linux-headers-6.1.0-11-common/arch/x86/include/asm/ptrace.h:97,
                 from /usr/src/linux-headers-6.1.0-11-common/arch/x86/include/asm/math_emu.h:5,
                 from /usr/src/linux-headers-6.1.0-11-common/arch/x86/include/asm/processor.h:13,
                 from /usr/src/linux-headers-6.1.0-11-common/arch/x86/include/asm/timex.h:5,
                 from /usr/src/linux-headers-6.1.0-11-common/include/linux/timex.h:67,
                 from /usr/src/linux-headers-6.1.0-11-common/include/linux/time32.h:13,
                 from /usr/src/linux-headers-6.1.0-11-common/include/linux/time.h:60,
                 from /usr/src/linux-headers-6.1.0-11-common/include/linux/stat.h:19,
                 from /usr/src/linux-headers-6.1.0-11-common/include/linux/module.h:13,
                 from //usr/include/wanpipe/wanpipe_includes.h:247,
                 from /usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:125:
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c: In function ‘sdla_busdma_map’:
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:11010:37: error: implicit declaration of function ‘pci_map_single’; did you mean ‘dma_map_single’? [-Werror=implicit-function-declaration]
11010 |                         cpu_to_le32(pci_map_single(hwcard->u_pci.pci_dev,dma_descr->dma_virt,map_len,dir));
      |                                     ^~~~~~~~~~~~~~
/usr/src/linux-headers-6.1.0-11-common/include/uapi/linux/byteorder/little_endian.h:34:51: note: in definition of macro ‘__cpu_to_le32’
   34 | #define __cpu_to_le32(x) ((__force __le32)(__u32)(x))
      |                                                   ^
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:11010:25: note: in expansion of macro ‘cpu_to_le32’
11010 |                         cpu_to_le32(pci_map_single(hwcard->u_pci.pci_dev,dma_descr->dma_virt,map_len,dir));
      |                         ^~~~~~~~~~~
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c: In function ‘sdla_busdma_unmap’:
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:11089:17: error: implicit declaration of function ‘pci_unmap_single’; did you mean ‘dma_unmap_single’? [-Werror=implicit-function-declaration]
11089 |                 pci_unmap_single(       hwcard->u_pci.pci_dev,
      |                 ^~~~~~~~~~~~~~~~
      |                 dma_unmap_single
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c: In function ‘sdla_busdma_sync’:
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:11142:20: error: ‘PCI_DMA_TODEVICE’ undeclared (first use in this function); did you mean ‘DMA_TO_DEVICE’?
11142 |         if (dir == PCI_DMA_TODEVICE) {
      |                    ^~~~~~~~~~~~~~~~
      |                    DMA_TO_DEVICE
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:11142:20: note: each undeclared identifier is reported only once for each function it appears in
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:11143:9: error: implicit declaration of function ‘pci_dma_sync_single_for_device’; did you mean ‘dma_sync_single_for_device’? [-Werror=implicit-function-declaration]
11143 |         pci_dma_sync_single_for_device(hwcard->u_pci.pci_dev, dma_descr->dma_addr,
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         dma_sync_single_for_device
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:11146:17: error: implicit declaration of function ‘pci_dma_sync_single_for_cpu’; did you mean ‘dma_sync_single_for_cpu’? [-Werror=implicit-function-declaration]
11146 |                 pci_dma_sync_single_for_cpu(hwcard->u_pci.pci_dev, dma_descr->dma_addr,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                 dma_sync_single_for_cpu
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c: In function ‘sdla_register’:
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:5166:20: warning: this statement may fall through [-Wimplicit-fallthrough=]
 5166 |                 if (conf != NULL){
      |                    ^
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:5169:9: note: here
 5169 |         case WANOPT_AFT:
      |         ^~~~
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c: In function ‘sdla_down’:
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:6565:28: warning: this statement may fall through [-Wimplicit-fallthrough=]
 6565 |                         if (hwcpu->used > 1){
      |                            ^
/usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.c:6569:17: note: here
 6569 |                 default:
      |                 ^~~~~~~
cc1: some warnings being treated as errors
make[2]: *** [/usr/src/linux-headers-6.1.0-11-common/scripts/Makefile.build:255: /usr/src/wanpipe-7.0.34/kdrvtmp/sdladrv_src.o] Error 1
make[1]: *** [/usr/src/linux-headers-6.1.0-11-common/Makefile:2037: /usr/src/wanpipe-7.0.34/kdrvtmp] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-6.1.0-11-amd64'
make: *** [Makefile:33: all] Error 2

---------------------------------------------

```

**Comment by InterLinked1** [comment]
This part is new to me. Can you include that in your support ticket?
Unfortunately, wanpipe is not a Sangoma open source project (it is open source, but it isn't a community open source project like Asterisk or DAHDI). So it is entirely on them to fix this issue, and they need to know it doesn't work.
Since you have a Sangoma card, make sure to include that. You are entitled to support from Sangoma in getting this fixed, so best to make full use of that and pester them until it's fixed. They've been very sloppy with this for far too long.

**Comment by Famicoman** [comment]
I went ahead and submitted an inquiry with sales, but when I load the support site, the right panel where I assume there is some contact info or a form is not populating. This is happening on all my machines and browsers. Are you experiencing this as well? It seems I have no way to actually contact or figure out how to contact support.

**Comment by Famicoman** [comment]
Okay I was able to create a support ticket. Below is the body of the message I sent. Excuse the formatting, they only accept plaintext:

```
Description
I have a new server running stock Debian 12 and am trying to install the newest version of the Wanpipe driver (7.0.35) in order to support my Sangoma A101 card but am hitting two issues.

1) It appears that the Setup script does not properly support Debian and fails with error: 
./Setup: line 2654: [: -eq: unary operator expected
ERROR: Unsupported Kernel Version 6.1.0-11-amd64
This appears to be because of the function get_kernel_ver  looking for VERSION/PATCHLEVEL/SUBLEVEL/EXTRAVERSION in the location /lib/modules/6.1.0-11-amd64/build/Makefile on my system, however this file on my system only contains an #include: 
# cat /lib/modules/6.1.0-11-amd64/build/Makefile
include /usr/src/linux-headers-6.1.0-11-common/Makefile
If I check the file at /usr/src/linux-headers-6.1.0-11-common/Makefile I see the proper values Setup is expecting.

2) Provided the above points to the proper directory, which I have done via a workaround on my local system by modifying Setup to perform every call to get_kernel_ver with the directory:
get_kernel_ver /usr/src/linux-headers-6.1.0-11-common
I now see that it is still failing with message ERROR: Unsupported Kernel Version 6.1.0-11-amd64
This is due to there being no logic in the file for kernel 6.X. For example I see logic for kernels 3.x, 4.x, and 5.x in this line:  if [ "$KVER" -eq "3" ] || [ "$KVER" -eq "4" ] || [ "$KVER" -eq "5" ]; then, but no check for "6".
```

**Comment by InterLinked1** [comment]
Awesome, let us know what you hear back.

I wouldn't hesitate to be aggressive if needed. Since you have a Sangoma card which is under warranty, since it essentially is unusable currently, and they have known about this now for quite some time, if they are not able to provide a fix within a few days, I would ask for a complimentary Digium card that doesn't require wanpipe.

**Comment by Famicoman** [comment]
Support stated that their expectations at the moment are that Debian 12/ Kernel 6.X support is lagging but expect that this will be enhanced later this year. I was asked if I had a particular business case for using Debian 12, and frankly I don't as I only installed it since it was the latest.

I decided to redo my system with Debian 11, as this really makes more sense since it is an LTS release but still had an issue in that the wanpipe module would compile but not load properly. 

I got really fed up and tried an alternate build via Osmocom's wanpipe notes: https://osmocom.org/projects/retronetworking/wiki/Sangoma_Wanpipe_DAHDI,

```
# git clone https://gitea.osmocom.org/retronetworking/wanpipe/
# cd wanpipe
# make
# make install
# modprobe wanpipe
# modprobe dahdi
# modprobe wanec
# dmesg
...
[ 1309.460950] WANPIPE(tm) Interface Support Module 7.0.34.0 (c) 1994-2016 Sangoma Technologies Inc
[ 1309.475605] WANPIPE(tm) Multi-Protocol WAN Driver Module 7.0.34.0 (c) 1994-2016 Sangoma Technologies Inc
[ 1309.475607] wanpipe: Probing for WANPIPE hardware.
[ 1309.475936] wanpipe: AFT-A101-SH PCIe T1/E1 card found (HDLC (DS) rev.39), cpu(s) 1, bus #3, slot #4, irq #11
[ 1309.475942] wanpipe: Allocating maximum 1 devices: wanpipe1 - wanpipe1.
[ 1309.476125] WANPIPE: TDM Codecs Initialized
[ 1359.550691] WANPIPE(tm) WANEC Layer 7.0.34.0 (c) 1995-2006 Sangoma Technologies Inc.
[ 1359.550693] wanec_create_dev: Registering Wanpipe ECDEV Device!
# cat /proc/net/wanrouter/hwprobe
-------------------------------
| Wanpipe Hardware Probe Info |
-------------------------------
1 . AFT-A101-SH : SLOT=4 : BUS=3 : IRQ=11 : CPU=A : PORT=1 : HWEC=32 : V=39

Sangoma Card Count: A101-2=1
```

Looks to be working based on the above, though I haven't tested connecting the channel bank up yet.

I suppose I could keep following up with Sangoma regarding Debian 11 / Kernel 5, but I imagine there would be more back-and forth and I'd have to undo my (hopefully) working system to follow through with them.

**Comment by InterLinked1** [comment]
Honestly, I think they need to meet their users where they are, if somebody wants to use Debian 12 they should be able to. A reason you could use is being able to use the newest packages, e.g. the newest version of `gcc` compiler. That's a big impetus for my ongoing upgrades right now. So you could follow up and say, actually, yeah, Debian 12 is kind of needed, just to ensure they keep working on it. Otherwise, they might never get to it.

Demanding a full refund or exchange of your card if they're not able to support Debian 12 by the end of the month could be a good place to start. These are based on internal tips I've gotten from Sangoma support. You need to really hound support and say you're going to walk away, or they just do diddly squat.

I wouldn't give them very much slack, again they've had almost a year to fix the other issue and haven't, which I think is downright unacceptable.

**Comment by InterLinked1** [comment]
@Famicoman So apparently we were one version behind on wanpipe, since there was a release with no announcement. If you run `phreaknet update` and try again, without further modifications, I *think* it should work now, on Debian 11. As you said they are still working on it for Debian 12.

**Comment by InterLinked1** [comment]
This appears to be fixed as of wanpipe 7.0.37.1, which was released in late December, and seems to work fully now on Debian 12 with 6.x kernels, so I'm closing this for now - feel free to reopen if you're still having issues!
